pkg/requester/selection/all.go

```diff
@@ -0,0 +1,139 @@
+}
+
+// The AllNodeSelector implements a node selection algorithm that executes the
+// job on all nodes that meet the job requirements.
+func NewAllNodeSelector(params AllNodeSelectorParams) requester.NodeSelector {
+	return &allNodeSelector{nodeDiscoverer: params.NodeDiscoverer, nodeRanker: params.NodeRanker}
+}
+
+// We want to select all nodes that match the job requirements. This will be any
+// node that passes node ranking with a positive value. We assume from this
+// point onwards that there is >= 1 execution for each node that passed the node
+// rank so that we don't have to keep track of which nodes passed in a store.
+func (s *allNodeSelector) SelectNodes(ctx context.Context, job *model.Job) ([]model.NodeInfo, error) {
+	ctx = log.Ctx(ctx).With().Str("JobID", job.ID()).Logger().WithContext(ctx)
+
+	nodeIDs, err := s.nodeDiscoverer.FindNodes(ctx, *job)
+	if err != nil {
+		return nil, err
+	}
+	log.Ctx(ctx).Debug().Int("Discovered", len(nodeIDs)).Msg("Found nodes for job")
+
+	rankedNodes, err := s.nodeRanker.RankNodes(ctx, *job, nodeIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	// filter nodes with rank below 0
+	filteredNodes := lo.Filter(rankedNodes, func(node requester.NodeRank, _ int) bool { return node.Rank >= 0 })
+	log.Ctx(ctx).Debug().Int("Ranked", len(filteredNodes)).Msg("Ranked nodes for job")
+
+	selectedInfos := lo.Map(filteredNodes, func(node requester.NodeRank, _ int) model.NodeInfo { return node.NodeInfo })
+	return selectedInfos, nil
+}
+
+// We asked all nodes that should be capable of running the job to bid, so we
+// want to select all the nodes that said they would run the job.
+func (*allNodeSelector) SelectBids(ctx context.Context, job *model.Job, jobState *model.JobState) (accept, reject []model.ExecutionState) {
+	return jobState.GroupExecutionsByState()[model.ExecutionStateAskForBidAccepted], nil
+}
+
+// We specifically want to run the job on *all* nodes – so we need to only retry
+// executions on the *same* node on which it failed. There would be no point /
+// danger in scheduling the execution to a different node.
+//
+// So we just select all of the nodes that have *only* failed executions (a node
+// that failed and then retried successfully should not be retried again!). This
+// only helps work around transient errors (e.g. something random wrong at the
+// node) and not permanent failures, so there is a limit to the number of times
+// we should retry.
+func (s *allNodeSelector) SelectNodesForRetry(ctx context.Context, job *model.Job, jobState *model.JobState) ([]model.NodeInfo, error) {
+	// failedNodes counts the number of failures that the node has seen so far.
+	// A negative value means the job is successful or in progress.
+	failureCounts := make(map[string]int, len(jobState.Executions))
+	for _, execution := range jobState.Executions {
+		numFails, seen := failureCounts[execution.NodeID]
+		if !seen {
+			numFails = 0
+		}
+
+		if numFails < 0 || !execution.State.IsDiscarded() {
```

Comment:
Would `IsActive` be more appropriate than `!IsDiscarded`? `IsActive` also includes the `ExecutionStateCompleted` state.